### ISSUE TITLE
calling this.render inside beforender causes nanocomponent to not unmount when necessary

### ISSUE THREAD

**Issue by lrlna** [choojs/nanocomponent#51]
I've experienced an interesting bug where I need to use `beforerender` to make an xhr call to get some data, and change my component based on the data i get in xhr call, after the call is complete, i have been calling `this.render()` to re-render the element with the new data. This is an example of my component:

```js
PreviewComponent.prototype.createElement = function (state, emit) {
  this.state = state
  this.emit = emit

  return html`
    <div class="container">
      <div id="editor-container" class="row">
        ${this.preview}
      </div>
    </div>
  `
}

PreviewComponent.prototype.beforerender = function () {
  // make the xhr request and get data back
  this._getScript((data) => {
    const html = document.createElement('div')
    html.innerHTML = data.html

    this.preview = html
    // this bugs out!
    this.render(this.state, this.emit)
  })
}
```

I was trying to log see the component on create, and what would happen when i theoretically would unload the component. The component doesn't unload, but does come back with the element being empty. i.e.:
<img width="625" alt="screen shot 2017-08-30 at 19 48 55" src="https://user-images.githubusercontent.com/8107784/29887954-6b713a1c-8dbf-11e7-97e6-614955d814f1.png">

I am currently just hacking around it and instead of render just calling a set timeout, which helps with unmountinng the component for now; something like this:
```
setTimeout(function () {
  // render
}, 0)
```

I would make a PR but I am still not sure how to fix it, if anyone knows how, lmk! 

Thanks ^__^

**Comment by yoshuawuyts** [comment]
This seems pretty severe - don't think there's a way to guard people from
calling `this.render()` within `.beforeload`. At best we could do is
document that folks shouldn't be doing it, but it won't guard - and if they
still do it they'll probably run into a bug that's hard to debug without
deep knowledge of the system.

Any thoughts what to do here?

**Comment by bcomnes** [comment]
```
<lrlna> bret: yea if before render is meant to be used to only modify or attach, then i am definitely using it wrong, i am basically populating the component
11:54 so yea, putting it in load makes a ton more sense
11:55 <yoshuawuyts> bret: replied to the issue
11:55 bret: think that bug is bad news for our API
11:55 bret: but luckily not a bug within the actual implementation
11:55 bret: we should probably give some thought on how to deal with it tho
11:56 <bret> Bret Comnes yoshuawuyts: I think maybe having an internal bus that people can emit('render') 
11:56 and we can put up guards or warnings if people are trying to internally re-render before mount
11:57 essentially introducing a new interface external rendering vs internal rendering
11:57 right now internal rendering is a hack using external rendering alls
11:57 <yoshuawuyts> bret: !!!
11:57 bret: I dig that idea
11:58 <bret> Bret Comnes we would probably need to cache arguments again
11:58 <yoshuawuyts> bret: I'm all about putting up heaps of guards
11:58 
```

**Comment by ungoldman** [comment]
I'm for caching args anyway 👍

**Comment by bcomnes** [comment]
```
<bret> Bret Comnes yoshuawuyts: yeah
13:26 also I had an idea for this thing
13:26 <yoshuawuyts> bret: we need a complimentary emoji for https://github.com/choojs/awesome-choo
13:26 :D
13:26 your selection so far has been A++
13:26 <bret> Bret Comnes this.rerender()
13:26 same as .render() but apply's last args
13:27 <yoshuawuyts> oh yeah, that could work
13:27 <bret> Bret Comnes easy to add
13:27 straightforward
13:27  ¯\_(ツ)_/¯ 
13:27 <yoshuawuyts> yeah, also makes .render() less magicky
13:27 if we were to cache args in there
13:27 errr
```

**Comment by lrlna** [comment]
@bcomnes @yoshuawuyts hmm yea `this.rerender()` would be a lot more straight forward. I think i'd think about it a bit more when I want to trigger it and consciously have to make a choice about re-rendering an element.  

But definitely making a documentation note on not changing an element in `beforerender` would totally be good right now.

**Comment by bcomnes** [comment]
https://github.com/choojs/nanocomponent/pull/52

I'll update the docs too

**Comment by lrlna** [comment]
oh cool! didn't see the PR o.O

**Comment by ungoldman** [comment]
Checking up on old issues -- @lrlna was this for you resolved by #52?

**Comment by lrlna** [comment]
@ungoldman hmmm i think that solved it for me, yea. I don't work on that codebase anymore so i am not certain. I say let's close it until we know otherwise?
